### ISSUE TITLE
PAGE checks do not work when the "external_modules" development version of the module framework is present

### ISSUE THREAD

**Issue by mmcev106** [Research-IT-Swiss-TPH/redcap-pdf-injector#35]
@tertek, specifically the `PAGE === "ExternalModules/index.php"` check does not work as expected.  I'm scheming the best recommendation to fix this...  I wonder if updating the `$module->isPage()` method as follows is the most intuitive:

> isPage([$path]) | 9.7.6 | Prior to framework version 14, `$path` is required and `true` is returned if the current page matches a path under the REDCap directory (e.g. `Design/online_designer.php`).  For framework version 14 and later, `$path` is optional.  If specified, `true` is also returned if `$path` matches the name/path for a module page (e.g. `some-module-page`).  If `$path` is not specified, `true` is returned if any page for the current module is being displayed.

What do you think?  Do have have any alternative ideas?

**Comment by tertek** [comment]
@mmcev106  I am not sure if  handling this with `$module->isPage()` is the most intuitive.
In my opinion, a better approach could be the introduction of a new method: `isModulePage`:

> isModulePage([$path]) | From framework version 14 and later , `$path` is optional. If specified, `true` is returned if  `$path` matches the name for a module path/page (e.g. some-module-page). If $pathis not specified, true is returned if any page for the current module is being displayed.

This would also be practical, since currently `$module->isPage` used `APP_PATH_WEBROOT` that breaks as well when "external_modules" development version of the module framework is present.

I could not find it, but is there any flag/constant set, in case "external_modules" development version is active/present?
In that way it would be much easier to handle the case.

**Comment by tertek** [comment]
Some quick write up:

```php
        /**
	 * @return bool
	 */
	function isModulePage($path=""){

        $url_prefix = "?prefix=" . $this->getPrefix();

        $url_extmod_web = APP_PATH_WEBROOT . "ExternalModules/" . $url_prefix;
        $url_extmod_dev = APP_URL_EXTMOD_RELATIVE . $url_prefix;

        //  check if URI contains extmod url and module prefix
        $check_web = strpos($_SERVER['REQUEST_URI'], $url_extmod_web) !== false;
        $check_dev = strpos($_SERVER['REQUEST_URI'], $url_extmod_dev) !== false;

        //  return false if not a module page at all
        if(  !$check_web && !$check_dev ) {
            return false;
        }

        //  return true if no specific page has been defined
        if($path === "") {
            return true;
        }

        //  return true if url parameter for p
        if($_GET["page"] == $path) {
            return true;
        }

        return false;
        
    }
```

**Comment by mmcev106** [comment]
@tertek, I'm now thinking of adding an `isModulePage()` method as you suggested, but also effectively renaming `isPage()` to `isREDCapPage()` (while continuing to support existing `isPage()` calls).  I'm also considering adding javascript versions of each method.  Do you have any more thoughts on this before I do a community post about it?

**Comment by tertek** [comment]
Great news!Will return to you after Tuesday. Thank you for keeping me in the loop. -------- Original Message --------From: Mark McEver ***@***.***>Date: Fri, May 26, 2023 6:36 PM +0200To: Research-IT-Swiss-TPH/redcap-pdf-injector ***@***.***>CC: Ekin Tertemiz ***@***.***>, Mention ***@***.***>Subject: Re: [Research-IT-Swiss-TPH/redcap-pdf-injector] PAGE checks do not work when the "external_modules" development version of the module framework is present (Issue #35)
@tertek, I'm now thinking of adding an isModulePage() method as you suggested, but also effectively renaming isPage() to isREDCapPage() (while continuing to support existing isPage() calls).  I'm also considering adding javascript versions of each method.  Do you have any more thoughts on this before I do a community post about it?

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>
[
{
***@***.***": "<a href="http://scanmail.trustwave.com/?c=22737&amp;d=kODw5EWk9avanE0nP_2oI5c_8Atoe0XmEy8NlAt4fQ&amp;u=http%3a%2f%2fschema%2eorg">http://schema.org</a>",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "<a href="https://scanmail.trustwave.com/?c=22737&amp;d=kODw5EWk9avanE0nP_2oI5c_8Atoe0XmEy8Fl18ldw&amp;u=https%3a%2f%2fgithub%2ecom%2fResearch-IT-Swiss-TPH%2fredcap-pdf-injector%2fissues%2f35%23issuecomment-1564647386">https://github.com/Research-IT-Swiss-TPH/redcap-pdf-injector/issues/35#issuecomment-1564647386</a>",
"url": "<a href="https://scanmail.trustwave.com/?c=22737&amp;d=kODw5EWk9avanE0nP_2oI5c_8Atoe0XmEy8Fl18ldw&amp;u=https%3a%2f%2fgithub%2ecom%2fResearch-IT-Swiss-TPH%2fredcap-pdf-injector%2fissues%2f35%23issuecomment-1564647386">https://github.com/Research-IT-Swiss-TPH/redcap-pdf-injector/issues/35#issuecomment-1564647386</a>",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "<a href="https://scanmail.trustwave.com/?c=22737&amp;d=kODw5EWk9avanE0nP_2oI5c_8Atoe0XmE3lcmwpwKg&amp;u=https%3a%2f%2fgithub%2ecom">https://github.com</a>"
}
}
]

**Comment by tertek** [comment]
I think the separation into `isREDCapPage()` and `isModulePage()` is very good.
In the future, they might support checks such as $isAuthenticated or $hasRole.
Also the Javascript versions are very neat!

**Comment by mmcev106** [comment]
@tertek, I just created a [PR adding isREDCapPage() and isModulePage() methods](https://github.com/vanderbilt-redcap/external-module-framework/pull/602).  I opted to pass on the JS versions, as they would add complication, and it is easy to JSON encode any PHP calls needed for use in JS.  Don't hesitate to speak up if you have any thoughts on this PR, or would like more than a few days to consider it.
